File: disturbances/src/main/java/md/vnastasi/cloud/client/model/DisturbanceWrapper.java
package md.vnastasi.cloud.client.model;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", visible = true)
@JsonSubTypes({
        @JsonSubTypes.Type(value = CalamityWrapper.class, name = "CALAMITY"),
        @JsonSubTypes.Type(value = DisruptionWrapper.class, names = {"MAINTENANCE", "DISRUPTION"}),
})
public sealed interface DisturbanceWrapper permits CalamityWrapper, DisruptionWrapper {

    String id();

    DisturbanceTypeWrapper type();

    String title();

    boolean isActive();
}

File: disruptions/src/main/kotlin/md/vnastasi/cloud/client/model/disruption/DisturbanceWrapper.kt
package md.vnastasi.cloud.client.model.disruption

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty
import java.time.OffsetDateTime

data class DisturbanceWrapper(

    @JsonProperty("id")
    val id: String,

    @JsonProperty("type")
    val type: DisturbanceTypeWrapper,

    @JsonProperty("header")
    val header: String,

    @JsonProperty("verwachting")
    val expectation: String?,

    @JsonProperty("oorzaak")
    val cause: String?,

    @JsonProperty("alternatiefVervoer")
    val alternativeTransport: String?,

    @JsonProperty("meldtijd")
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ssX")
    val reportedAt: OffsetDateTime?,

    @JsonProperty("trajecten")
    val trajectories: List<TrajectoryWrapper>?,

    @JsonProperty("extraReistijd")
    val extraTravelTime: String?,

    @JsonProperty("reisadviezen")
    val travelAdvice: TravelAdviceWrapper?,

    @JsonProperty("geldigheidsLijst")
    val validityList: List<ValidityWrapper>?,

    @JsonProperty("gevolg")
    val consequence: String?,

    @JsonProperty("periode")
    val period: String?
)
